Implement return water reset setpoint manager

Myoldmopar opened the pull request:
[#1836664]

This still has work like:
- [x] Get IDD comments to get better names
- [x] Write unit tests
- [x] Write docs

But frankly, it does work great:
- [Living design document](https://docs.google.com/document/d/1YKcKymPZwCODDoQS6PTLg--F5YSb-F4HHqIH4dJx6Tg/edit?usp=sharing)

Follow-up work from review call:
- [X] Rename objects
  - SetpointManager:SupplyResetForReturnTemperature:ChilledWater -> SetpointManager:ReturnTemperature:ChilledWater
  - SetpointManager:SupplyResetForReturnTemperature:HotWater -> SetpointManager:ReturnTemperature:HotWater
  - Completed in idd, example files, source code, and docs
- [X] Collapse three example files into one.  I'm going to scratch this given that some real investigation was requested, and that seems more important for my time.  If there is a strong push to get them collapsed, I'm OK with doing it, but for now I'm marking it complete (scratched)
- [x] Add the ability to mine the plant loop data structure to get the actual setpoint node on the plant in case it isn't the supply outlet node.
- [x] Change field names:
  - Take out design everywhere
  - Make the 'design' setpoint as the min/max limit, make sure language is clear
  - Add setpoint or something to the return temperature "setpoint"
- [x] Add the ability to read the current return temperature target from the actual `Node.TempSetPoint` value on the return node that could be set by any other method.  Need to check for -999 and warn though.
- [X] Investigate calling the setpoint manager more frequently from plant, isolate only plant setpoints...which ones?
  - Added in a special call inside the plant manager at the beginning of a set of plant iterations.  Verified it was being called.  It didn't make hardly any difference.  I won't spend any more time on this, especially after hooking it up to the large office model with the current calling frequency.
- [X] Simulate with a simple office building model and look at the response with a wide time step/short time step
  - Looked at it in current state with the new setpoint manager [here](https://github.com/NREL/EnergyPlus/pull/4909#issuecomment-102130725) and it looks great.

Myoldmopar committed:
Implement return water reset setpoint manager;
[#1836664]

Myoldmopar committed:
Initial unit test implementation, will flush out tomorrow

Myoldmopar committed:
Completed unit test implementation

Myoldmopar committed:
IORef for setpoint managers

Myoldmopar committed:
Add snippets to IORef, Finish EngRef

Myoldmopar commented:
OK, everything is flushed out, including unit tests and docs.  Maybe I can hijack a few minutes of the conference call today to go over naming, because I am thoroughly aware that the naming can use work.  From that outcome I'll make final changes, then I think this should be ready to merge rather quickly.  

(FYI @mjwitte)

Myoldmopar commented:
@mjwitte @wfbuhl @EnergyArchmage @rraustad I tried to recreate the comments/suggestions from the review meeting into the [initial pull request comment](https://github.com/NREL/EnergyPlus/pull/4909#issue-75749940) above.  If you think of anything I missed, just leave it in a comment and I'll add it above.

Myoldmopar commented:
Took the RefBldgLargeOffice model and simply swapped out the scheduled setpoint manager for CoolSys1 for the new model.  Here is the response from before:
![image](https://cloud.githubusercontent.com/assets/849824/7638945/b9750e0c-fa35-11e4-830c-fe10f42a5b06.png)

During part load conditions, such as in the morning, since the supply setpoint is fixed, the return temperature floats.  All good.  Now here is with the return temperature targeted at 12 degrees fixed:
![image](https://cloud.githubusercontent.com/assets/849824/7638950/c9000566-fa35-11e4-9c12-5f88036444c8.png)

Note that during the part load morning conditions, the supply setpoint increases to accommodate the smaller delta T.  Also note that the morning pick up load causes the delta T to go too high, so the return temperature does go above target, but then steadies back out.

### NOTE

This is with the setpoint manager called exactly as it is now, no higher frequency.  Given this very nice response, I'm not going to investigate the calling frequency any further.

(FYI @mjwitte @EnergyArchmage)

mjwitte commented:
Looks good.

Myoldmopar committed:
Renaming object and field names, re-order hot-water for alignment, other cleanup

EnergyArchmage commented:
Nothing like hourly reporting to smooth things out.  Looks great.

Myoldmopar commented:
And much less drastic step changes in loop demand...

Myoldmopar committed:
Mined plant data structure for actual setpoint node

Myoldmopar committed:
Add the option to lookup the target return temperature;
This is looked up on the return node TempSetPoint.;
Included another example file, and updated docs

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 1836664-ReturnWaterTempReset

Myoldmopar commented:
@EnergyArchmage @mjwitte I've addressed everything that came up in the conference call yesterday.  I merged in develop and pushed up the latest changes.  CI shouldn't have any problem with it.

I'm happy to go over the changes with you, though I think I'm going to give @kbenne a heads up about this new feature as he will likely have interest in it.  He might be willing to merge it in once I go over the feature and testing with him.

You have anything else you'd like from me?  Or would you give your blessing :pray: for this merge :wedding: ?

Myoldmopar commented:
The RefBldg diffs were accidental just due to me trimming outputs for convenience.  I've reverted it back and it should come back fully clean now.

Myoldmopar committed:
Revert the temporary changes to the large office building model

rraustad commented:
I've only a feeling that this setpoint manager should track plant loop iterations. What other issues that control might reveal is to be determined. Yes, I think this should be merged in due to it's 1) accurate tracking of return water temp and, 2) the manufacturers request for this feature. Optimization of this feature will follow in due time.

Myoldmopar commented:
Thanks @rraustad I have other tasks in this same part of E+, so I'll try to throw in additions as I go.

rraustad commented:
Need more work? Did you try this on a condenser loop? I would expect the same result.

rraustad commented:
"Add the ability to mine the plant loop data structure to get the actual setpoint node on the plant in case it isn't the supply outlet node."

If this was completed, then could we instead just ask for a plant loop name instead of inlet/outlet node? Just thinking about future transitions if we move in this direction...yes I do know there's an issue here with available data.
-  A2,  \field Plant Loop Supply Outlet Node
-  A3,  \field Plant Loop Supply Inlet Node

Myoldmopar commented:
Yeah eventually we should be able to all for a pant loop name. And the transition will be easy to write if we make that change.

Myoldmopar commented:
Oh and it will work on condenser loops.

EnergyArchmage commented:
I guess I was thinking all new features had be OO and need to make some effort to refactor code  touched to go in that direction. This seems to be extending the same old pattern and makes the policy unclear to me.

Myoldmopar commented:
This is object oriented.  Object-oriented doesn't require a specific attribute like inheritance, although that is definitely something to consider.  The structs I added include an instance method, which is almost entirely black-boxed such that it doesn't rely on _almost_ any global data to be set up to make it easily testable.  (PlantLoop should be set up so that it can mine the setpoint node.)

I have other tasks in this area to update the condenser setpoint reset for multiple towers, etc., so I'll push that stuff a little farther as I go.

Myoldmopar committed:
Final updates to docs with updated info from latest testing/changes

Myoldmopar commented:
@mjwitte Are you willing/able to be assigned to this?  I made some changes to the docs and am pushing them now.  I think it should be ready to merge unless you or others have any additional concerns.

mjwitte commented:
@Myoldmopar Sure, I can review, but it won't be until Monday.

Myoldmopar commented:
Thanks @mjwitte, Monday will be awesome.  I assigned the Pivotal tickets to you.

mjwitte commented on testfiles/PlantLoadProfileCooling.idf:
```diff
@@ -0,0 +1,398 @@
+
+  Output:Variable,*,Plant Load Profile Heat Transfer Energy,Hourly;
+
+  Output:Variable,*,District Cooling Hot Water Rate,Hourly;
```
@Myoldmopar This output variable and the next should say "Chilled Water" instead of "Hot Water".  I was expecting PlantLoadProfileCooling.idf and PlantLoadProfileCoolingReturnReset.idf to differ only in the chilled water loop setpoint controls.  Would you be open to redoing PantLoadProfileCooling by copying PlantLoadProfileCoolingReturnReset and simply changing the setpoint manager to supply temp control?

Myoldmopar replied:
Yeah, I can do that @mjwitte.  I'll push a commit here in a bit.

I guess this brings into question whether we want all the example files I added, though.  I am fine with trimming down the list.  Currently, I have in my branch:
- PlantLoadProfileCooling.idf: This is just a cooling version of PlantLoadProfile.  Not necessarily that useful.
- PlantLoadProfileCoolingReturnReset.idf: This is a load profile cooling plant but with the reset strategy enabled with a constant target return temperature. Useful.
- PlantLoadProfileCoolingReturnResetLookup.idf: This is a load profile cooling plant with the reset strategy enabled where it looks up the current target setpoint from the actual Node.TempSetPoint on the return node. Probably useful since it shows the secondary setpoint manager.
- PlantLoadProfileCoolingReturnResetVariableFlow.idf: This is the cooling plant with reset strategy, but with the load profile flow request varying through the day. Not especially useful.
- PlantLoadProfileHeatingReturnResetScheduled.idf: This is a heating plant with reset strategy that shows the target return temperature scheduled.  Not especially useful.
- ReturnReset-RefBldgLargeOfficeNew2004_Chicago.idf: This is the large office model with the reset strategy enabled.  Useful.

I'm happy to make sure they are all ready to go in, but I'm also fine trimming the list.  What do you think?

mjwitte commented:
@Myoldmopar If I were going to choose a short list, I'd probably keep PlantLoadProfileCoolingReturnResetVariableFlow and make PlantLoadProfileCooling based on that one (i.e. ditch the constant flow altogether).  And keep the PlantLoadProfileCoolingReturnResetLookup (and it should be variable flow as well then).
I think ReturnReset-RefBldgLargeOfficeNew2004_Chicago.idf is useful, but I would rename it to RefBldgLargeOfficeNew2004_Chicago-ReturnReset.idf  Or change the other ones to start with ReturnReset or - just seems odd to have only one of these start with ReturnReset.  I know we don't have a consistent pattern across the entire set of example files.

Myoldmopar committed:
Rename refbldg example and update CMakeLists accordingly.

Myoldmopar committed:
Remove scheduled heating file and update CMakeLists accordingly

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 1836664-ReturnWaterTempReset

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69827,6 +69827,110 @@ SetpointManager:SingleZone:OneStageHeating,
+       \type real
+       \required-field
+  A4,  \field Return Temperature Setpoint Input Type
+       \note This defines whether the chilled water return temperature target is constant, 
```
chilled --> hot

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69827,6 +69827,110 @@ SetpointManager:SingleZone:OneStageHeating,
+       \note supply setpoint.
+       \type node
+       \required-field
+  N1,  \field Minimum Supply Temperature Setpoint
```
All of the setpoint fields need units

mjwitte replied:
Also, defaults might be nice.

Myoldmopar commented:
OK, thanks Mike.  Here's what I've done for the test files:
- PlantLoadProfileCooling.idf: Hmm, I decided to just remove this.  It isn't really anything useful, more of just an interim file as I developed the other files.
- PlantLoadProfileCoolingReturnReset.idf: Made into variable flow.
- PlantLoadProfileCoolingReturnResetLookup.idf: Made into variable flow.
- PlantLoadProfileCoolingReturnResetVariableFlow.idf: Removed, it became redundant when the PlantLoadProfileCoolingReturnReset became variable flow.
- PlantLoadProfileHeatingReturnResetScheduled.idf: Removed
- ReturnReset-RefBldgLargeOfficeNew2004_Chicago.idf: Renamed and kept

I'll push these changes and update the other things you've noted shortly.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69827,6 +69827,110 @@ SetpointManager:SingleZone:OneStageHeating,
+  N3,  \field Return Temperature Setpoint Constant Value
+       \note This is the desired return temperature target, which is met by adjusting the  
+       \note supply temperature setpoint. This constant value is only used if 
+       \note the "Design Chilled Water Return Temperature Input Type" is Constant
```
The quotes are not needed on the field name reference, title-case is sufficient.

Myoldmopar committed:
Remove plain cooling load profile file, and update CMakeLists accordingly

Myoldmopar committed:
Remove now redundant variable-flow-specific file, update CMakeLists

Myoldmopar committed:
Address idd comments from @mjwitte

Myoldmopar committed:
Quick fix from goof on file move

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69827,6 +69827,110 @@ SetpointManager:SingleZone:OneStageHeating,
        \type node
 
+SetpointManager:ReturnTemperature:ChilledWater,
+     \memo This setpoint manager is used to place a temperature setpoint on a
```
This is a little repetitive and the first sentence says "fixed return . . ." and the second sentence says "which is scheduled."  How about something like (still a little clunky):
This setpoint manager is used to place a temperature setpoint on a plant supply outlet node based on a target return water setpoint. The setpoint manager attempts to achieve the desired return water temperature by adjusting the supply temperature setpoint based on the plant conditions at each system time step.

Myoldmopar commented:
Thanks for the review @mjwitte.  I believe I've addressed everything:
- Update files as per my above comment.  This resulted in three new files being added:
  - PlantLoadProfileCoolingReturnReset.idf, which demonstrates a fixed target return temperature with the variable flow and load plant profile
  - PlantLoadProfileCoolingReturnResetLookup.idf, which demonstrates the ability to use a secondary setpoint manager to specify the target return temperature
  - RefBldgLargeOfficeNew2004_Chicago-ReturnReset.idf, which demonstrates adding this control to a real building model
- Removed quotes from IDD, added units, added defaults, cleaned up wording

I'll address the comment you just made and give you a little more time before I push again in case you have anything else.

Myoldmopar committed:
Update IDD object memos

mjwitte commented:
Well, OK.  I thought PlantLoadProfileColing (with variable flow and standard supply temperature control) would be useful as a comparison case against the return reset, but that's not important.  

I hadn't intended for you to remove PlantLoadProfileHeatingReturnResetScheduled.idf  this leaves the HotWater version unused in any test files.  So, either need to restore that one, or perhaps use both the HotWater and ChilledWater return managers in the RefBldgLargeOffice one?

(Still need to read through the docs.  Will let you know when I'm finished.)

mjwitte commented on doc/src/docs/EngineeringReference/13e-EncyclopaedicRefs.md:
```diff
@@ -3121,6 +3121,64 @@ The object determines a “near-optimal” condenser water entering set point at
+
+#### Operation
+
+The setpoint senses conditions on two named nodes, which should correspond to a central plant's supply side inlet and outlet nodes, the return and supply nodes, respectively.  The temperature and flow rate information is used in a series of calculations in order to determine a setpoint to specify on the supply outlet node which will in turn provide the target desired return temperature on the sup ...
```
The setpoint _manager_ . . .

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -196,7 +244,9 @@ namespace SetPointManager {
 	int NumSZOneStageCoolingSetPtMgrs( 0 ); // number of single zone one stage cooling setpoint managers
 	int NumSZOneStageHeatingSetPtMgrs( 0 ); // number of singel zone one stage heating setpoint managers
-
+	int NumSZReturnWaterResetChWSetPtMgrs( 0 ); // number of return water reset setpoint managers
```
Minor quibble, this and next one don't need "SZ" in the variable name.

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -2679,6 +2761,120 @@ namespace SetPointManager {
+			// process the sense and actuate nodes
+			bool errFlag = false;
+			ReturnWaterResetChWSetPtMgr( SetPtMgrNum ).supplyNodeIndex = GetOnlySingleNode( cAlphaArgs( 2 ), errFlag, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Unknown, NodeConnectionType_SetPoint, 1, ObjectIsNotParent, cAlphaFieldNames( 2 ) ); // setpoint nodes
+			ReturnWaterResetChWSetPtMgr( SetPtMgrNum ).returnNodeIndex = GetOnlySingleNode( cAlphaArgs( 3 ), errFlag, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Unknown, NodeConnectionType_SetPoint, 1, ObjectIsNotParent, cAlphaFieldNames( 3 ) ); // setpoint nodes
```
NodeConnectionType_Sensor

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -2679,6 +2761,120 @@ namespace SetPointManager {
+			// process the sense and actuate nodes
+			bool errFlag = false;
+			ReturnWaterResetHWSetPtMgr( SetPtMgrNum ).supplyNodeIndex = GetOnlySingleNode( cAlphaArgs( 2 ), errFlag, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Unknown, NodeConnectionType_SetPoint, 1, ObjectIsNotParent, cAlphaFieldNames( 2 ) ); // setpoint nodes
+			ReturnWaterResetHWSetPtMgr( SetPtMgrNum ).returnNodeIndex = GetOnlySingleNode( cAlphaArgs( 3 ), errFlag, cCurrentModuleObject, cAlphaArgs( 1 ), NodeType_Unknown, NodeConnectionType_SetPoint, 1, ObjectIsNotParent, cAlphaFieldNames( 3 ) ); // setpoint nodes
```
NodeConnectionType_Sensor

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -6551,6 +6762,200 @@ namespace SetPointManager {
+
+		// we don't need fluid names since we have a real index, so just pass in the temperature and get properties
+		Real64 avgTemp = ( returnNode.Temp + supplyNode.Temp ) / 2;
+		Real64 cp = FluidProperties::GetSpecificHeatGlycol( "", avgTemp, fluidIndex, "ReturnWaterChWSetPointManager::calculate" );
```
ReturnWaterChWSetPointManager --> ReturnWaterHWSetPointManager

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -6551,6 +6762,200 @@ namespace SetPointManager {
+		if ( this->returnTemperatureScheduleIndex > 0 ) {
+			T_return_target = GetCurrentScheduleValue( this->returnTemperatureScheduleIndex );
+		} else if ( this->useReturnTempSetpoint ) {
+			if ( returnNode.TempSetPoint != -999 ) {
```
-999 should be SensedNodeFlagValue

mjwitte commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -6551,6 +6762,200 @@ namespace SetPointManager {
+				ShowContinueError( "The manager is specified to look to the return node setpoint to find a target return temperature, but the node setpoint was invalid" );
+				ShowContinueError( "Verify that a separate sepoint manager is specified to set the setpoint on the return node named \"" + NodeID( this->returnNodeIndex ) + "\"" );
+				ShowContinueError( "Or change the target return temperature input type to constant or scheduled" );
+			}
```
This should be fatal.  Other SPMs finish with
  ShowFatalError( "Missing reference setpoint." );

Myoldmopar committed:
Continued addressing review comments

Myoldmopar committed:
Add missing fatal error calls for missing return setpoint

Myoldmopar committed:
Update setpoint constant value instead of hardwire number

mjwitte commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -1980,6 +2056,9 @@ namespace SetPointManager {
 	UpdateSetPointManagers();
 
 	void
+	UpdatePlantSpecificSetPointManagers();
```
This doesn't appear to be used anywhere?

mjwitte commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -16285,9 +16289,143 @@ SetpointManager:SingleZone:OneStageHeating,
+
+#### Return Temperature Setpoint Constant Value
+
+If the "Design Hot Water Return Temperature Input Type" field is "constant", the numeric value entered here is used as the target return temperature throughout the simulation.
```
Old field name is lingering here "Design Hot Water Return Temperature Input Type" --> Return Temperature Setpoint Input Type.  Happens twice here and twice in CHW object.

Myoldmopar committed:
Remove stray definition from previous attempt to isolate plant setpoints

mjwitte commented:
I'm done with a pass over the docs and source.

One final concern, I don't see any validation of the node names to make sure they are on the same plant loop, or are even on any plant loop for that matter.  I tried putting a random node name (one from an air loop) in the Plant Loop Supply Inlet Node and the simulation ran without warnings and simply ended up riding one of the limits because the values coming back didn't make any sense.  And putting gibberish in the Plant Loop Supply Outlet Node results in this error, so I guess that's adequate:

```
   ** Severe  ** PlantManager: No Setpoint Manager Defined for Node=COOLSYS1 SUPPLY OUTLET NODE in PlantLoop=COOLSYS1
   **   ~~~   ** Add Temperature Setpoint Manager with Control Variable = "Temperature" for this PlantLoop.
   **  Fatal  ** Previous severe set point errors cause program termination
```

Perhaps you can check that both nodes are on the same plant loop when you find the plantIndex?

Myoldmopar committed:
Update old field names in docs

Myoldmopar commented:
@mjwitte I think this is now complete.  I added a convenience function to DataPlant to check whether two nodes are on the same plant loop.  I added a unit test to wrap this new function as well.

As long as CI likes this, and you don't have any additional review of the new stuff, this should be merge-able directly.  Thanks for your full review, it definitely made it a much more refined feature.

Myoldmopar committed:
Add validation of plant nodes for setpoint manager;
I added a new function to DataPlant for convenience;
Which then required a new unit test to be added for that function;
[#1836664]

Myoldmopar commented:
@mjwitte Of course, some of these may come back with the basement diff now that develop has moved.  I don't think it's worth re-merging develop back though.

Myoldmopar commented:
But I guess one more merge to get _all_ green wouldn't really hurt...

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 1836664-ReturnWaterTempReset

Myoldmopar commented:
@mjwitte Just a small nudge that I think I was able to address all of your requests to get this completed.  If CI comes back clean, I think it can be merged right away.

mjwitte committed:
Cleanup RefBldgLargeOfficeNew2004_Chicago-ReturnReset example file
